File: haskell/bob/Bob.hs
module Bob (responseFor) where

import Data.Char
import Data.List

removeSilience :: [Char]-> [Char]
removeSilience ws = [x| x <- ws, not( isSpace( x))]

removeNoAlpha :: [Char]-> [Char]
removeNoAlpha ws = [ x | x <- ws, isAlpha x]

responseFor :: [Char] -> [Char]
responseFor speak
  | null( removeSilience( speak)) = "Fine. Be that way!"
  | not( null( removeSilience( removeNoAlpha( speak)))) && all isUpper( removeNoAlpha( speak)) = "Woah, chill out!"
  | isSuffixOf "?" speak  = "Sure."
  | otherwise = "Whatever."

File: ruby/bob/bob.rb
class Bob

  def initialize
    @responses = {nothing: 'Fine. Be that way!',
                  yell:     'Woah, chill out!',
                  question:  'Sure.',
                  anything: 'Whatever.'}
  end

  def hey message
    answer message
  end

  private

  def answer message
    clean message

    @responses.select do |response, answer|
      send response, message
    end.values.first
  end

  def anything m; true end

  def question message
    message.end_with? '?'
  end

  def yell message
    message.match /^[^a-z]*$/
  end

  def nothing message
    message.empty?
  end

  def clean message
    message.strip!
  end
end
